eZ/Publish/Core/Persistence/Legacy/Content/Handler.php

```diff
@@ -321,12 +321,67 @@ public function load($id, $version, array $translations = null)
             $this->contentGateway->loadVersionedNameData(array(array('id' => $id, 'version' => $version)))
         );
         $content = $contentObjects[0];
+        unset($rows, $contentObjects);
 
         $this->fieldHandler->loadExternalFieldData($content);
 
         return $content;
     }
 
+    /**
+     * {@inheritdoc}
+     *
+     * @todo
+     * REVIEW NOTE: This methods ALWAYS handles AlwaysAvailable, this would perhaps be better handled in API layer like
+     *              done on load(), thus making it optional. AND faster as it can loadContentInfoList via SPI cache.
+     *              However then we would need a argument structure like used on gateway here.
+     *              Would that be OK for loadContentList() in SPI layer??
+     *              If so it should also take version number  as part of the structure.
```

Comment:
@alongosz & @Nattfarinn & @adamwojs : Sorry for this big ping, but: Any objections against having SPI method which has a bit strange argument? _(like exists in gateway today for named values, but also with translation)_

It will basically mean, instead of current:
`public function loadContentList(array $contentIds, array $translations): array`

It would become:
```php
    /**
     * @param array[] $contentFilter Array of arrays with keys `id`, `version` and `translations` 
     */
    public function loadContentList(array $contentFilter): array
```

I'm a bit torn on this myself, as it becomes one ugly ☠️  function, and as it moves some of the logic here from storage engine to consumers of this.

On the other side, the benefits are:
- It's just SPI.. 😈 
- It aligns with logic on `load()` where API today needs to load `ContentInfo` to pick currentVersion and append main language to languages to pick in case of always available flag
- It allows SPI cache to be reused between this method and `load()` as they then both knows the version number on what to lookup, so cache key can become the same.
- The content info lookup can be cached in SPI cache, which is not the case when we have to look it up here right now. That said, once cache is warmed it will mean 2 cache lookups instead of 1 _(But, unlike contentService->load() that is regardless of how many content id's we look for, so it does not grow)_
- It makes it possible to use this method, also for loading _not_ current versions., can even load several versions of same object if you'd like with that change.

I guess this ultimately boils down to where we want to take the SPI responsibility in the future, if we omit version numbers here, we should aim to allow it to be omitted from load() as well at some point and take advantage of that, _(or introduce a new function that does not need it)_.

I hope this was clear, feel free to reach out if something was not! ℹ️

Reply:
Overall, I like structs to load complicated items. They make BC handling easier, since we can have minor logic in there to handle new arguments. Side idea that came to me about those structs: since in the SPI layer we generate cache keys, may it be a good idea to have cache keys generated by these structs, like in `__toString()` ?

Reply:
That would be mixing layers, the struct being from SPI namespace knowns nothing about details of Persistent Cache, which is implementation details. On top of that it varies depending on cache lib used underneath.

Reply:
When it comes to usage of structs, I agree on any operations that change stuff, and while I'm open to it, be aware of:
- we haven't so far done it for load methods _(in SPI, API often have load by ContentInfo and such)_
- This is lookup by several items, so it would be array/collection of structs, so would be hard to model


Also be aware, if we _(somehow, in some kind of modeling)_ do what I propose above, API would be quite clear:
```php
public function loadContentList(ContentInfo[] $contentInfoList, array $translations): array
```

As the system would need to load content info anyway, so can just as well get consumer to do it to have proper clean API for this. So at least as far as API goes, it will not need any new struct for this.

Reply:
hmm... On further thinking in SPI, and also API for that matter, they can basically look like this:
```php
public function loadContentList(ContentInfo[] $list, array $translations, bool $useAlwaysAvailable): array
```

The secondary use case of loading several versions on same content can optionally also be exposed as the following if we ever need it:
```php
public function loadContentList(VersionInfo[] $list, array $translations, bool $useAlwaysAvailable): array
```

But for now I don't see a clear strong need for that, so this makes the method much more clear, and essentially no need to add new struct as we already have the right one for it.

WDYT @bdunogier / @Nattfarinn  / @alongosz / @adamwojs ?

Reply:
Reiterating usage of struct, we can for instance do the following, making sure we don't need to deal with alwaysAvailable logic in SPI which would be the case if we where to take ContentInfo as argument also on SPI. So SPI can be something like:
```php
public function loadContentList(LoadStruct[] $list): Content[]

// 

namespace eZ\Publish\SPI\Persistence\Content;

class LoadStruct  extends ValueObject
{
    public $id;
    public $version;
    public $prioritizedLanguages = [];
}
```



Then future API would be something like:
```php
public function loadContentList(ContentInfo[] $list, array $translations, bool $useAlwaysAvailable = true): Content[] | ContentCollection
```

Reply:
+1 for approach with LoadStruct. It looks much better than previous one :-)

Reply:
LoadStruct approach looks good but I don't understand the need for multi-typed return:
> ```php
> public function loadContentList(ContentInfo[] $list, array $translations, bool $useAlwaysAvailable = true): Content[] | ContentMap
> ```

We should never allow method returning multiple types. Something needs to be redesigned/refactored if this is needed.

Reply:
This was proposal for API, I'm not suggesting several types, just that the type returned is a enriched collection of Content items, so the pseudo code was me trying to explain that in shortest amount of text :)